pkg/apis/openebs.io/upgrade/v1alpha2/types.go

```diff
@@ -0,0 +1,195 @@
+/*
+Copyright 2019 The OpenEBS Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha2
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// +genclient
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+// +resource:path=upgradetask
+// +k8s:openapi-gen=true
+
+// UpgradeTask represents an upgrade task
+type UpgradeTask struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
+	// Spec i.e. specifications of the UpgradeTask
+	Spec UpgradeTaskSpec `json:"spec" protobuf:"bytes,2,opt,name=spec"`
+	// Status of UpgradeTask
+	Status UpgradeTaskStatus `json:"status" protobuf:"bytes,3,opt,name=status"`
+}
+
+// UpgradeTaskSpec is the properties of an upgrade task
+type UpgradeTaskSpec struct {
+	// FromVersion is the current version of the resource.
+	FromVersion string `json:"fromVersion" protobuf:"bytes,1,name=fromVersion"`
+	// ToVersion is the upgraded version of the resource. It should be same
+	// as the version of control plane components version.
+	ToVersion string `json:"toVersion" protobuf:"bytes,2,name=toVersion"`
+	// Flags contains the optional flags that can be passed during upgrade.
+	Flags Flags `json:"flags,omitempty" protobuf:"bytes,3,name=flags"`
+	// ResourceSpec contains the details of the resource that has to upgraded.
+	ResourceSpec `json:",inline" protobuf:"bytes,4,name=resourceType"`
+	// ImagePrefix contains the url prefix of the image url. This field is
+	// optional. If not present upgrade takes the previously present ImagePrefix.
+	ImagePrefix string `json:"imagePrefix" protobuf:"bytes,5,name=imagePrefix"`
+	// ImageTag contains the customized tag for ToVersion if any. This field is
+	// optional. If not present upgrade takes the ToVersion as the ImageTag
+	ImageTag string `json:"imageTag" protobuf:"bytes,6,name=imageTag"`
+}
```

Comment:
I guess all the properties define here needs to be understood by the corresponding image to proceed with upgrade activity.
In such a case, should we have `PodTemplate` or `JobTemplate` as a better representation?

Reply:
Would like to know the boundaries of upgrade executor versus k8s job which does the actual upgrade?
Does upgrade executor prohibit a volume upgrade job to be spawned if corresponding pool has not been upgraded?

Reply:
The upgrade job will fail with `csp deployment <csp-deployment-name> not in <upgrade-version> version`

Reply:
What is meant by corresponding image ? there will be only one image for the job.